openai-model-provider/proxy/proxy.go

```diff
@@ -96,6 +108,106 @@ func (s *server) proxyDirector(req *http.Request) {
+	if err := json.Unmarshal(bodyBytes, &reqBody); err == nil && isModelO1(reqBody.Model) {
+		modifyRequestBodyForO1(req, &reqBody)
+	} else {
+		req.Body = io.NopCloser(bytes.NewBuffer(bodyBytes))
+	}
+}
+
+func modifyRequestBodyForO1(req *http.Request, reqBody *openai.ChatCompletionRequest) {
+	reqBody.Stream = false
+	reqBody.Temperature = nil
+	for i, msg := range reqBody.Messages {
+		if msg.Role == "system" {
+			reqBody.Messages[i].Role = "developer"
+		}
+	}
+	modifiedBodyBytes, err := json.Marshal(reqBody)
+	if err == nil {
+		req.Body = io.NopCloser(bytes.NewBuffer(modifiedBodyBytes))
+		req.ContentLength = int64(len(modifiedBodyBytes))
+		req.Header.Set("Accept", "application/json")
+		req.Header.Set("Accept-Encoding", "")
+		req.Header.Set("Content-Type", "application/json")
+	} else {
+		fmt.Println("failed to marshal request body after modification and skipping, error: ", err.Error())
+	}
+}
+
+func (s *server) modifyResponse(resp *http.Response) error {
+	if resp.StatusCode != http.StatusOK || resp.Body == nil || resp.Request.URL.Path != chatCompletionsPath || resp.Request.URL.Host != openaiBaseHostName {
+		return nil
+	}
+
+	if resp.Header.Get("Content-Type") == "application/json" {
+		var respBody openai.ChatCompletionResponse
+		rawBody, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read response body: %w", err)
+		}
+		if err := json.Unmarshal(rawBody, &respBody); err == nil && isModelO1(respBody.Model) {
+			// Convert non-streaming response to a single SSE for o1 model
+			streamResponse := openai.ChatCompletionStreamResponse{
+				ID:      respBody.ID,
+				Object:  respBody.Object,
+				Created: respBody.Created,
+				Model:   respBody.Model,
+				Usage:   respBody.Usage,
+				Choices: func() []openai.ChatCompletionStreamChoice {
+					var choices []openai.ChatCompletionStreamChoice
+					for _, choice := range respBody.Choices {
+						choices = append(choices, openai.ChatCompletionStreamChoice{
+							Index: choice.Index,
+							Delta: openai.ChatCompletionStreamChoiceDelta{
+								Content:      choice.Message.Content,
+								Role:         choice.Message.Role,
+								FunctionCall: choice.Message.FunctionCall,
+								ToolCalls:    choice.Message.ToolCalls,
+							},
+							FinishReason: choice.FinishReason,
+						})
+					}
+					return choices
+				}(),
+			}
+
+			sseData, err := json.Marshal(streamResponse)
+			if err != nil {
+				return fmt.Errorf("failed to marshal stream response: %w", err)
+			}
+
+			sseFormattedData := fmt.Sprintf("data: %s\n\nevent: close\ndata: [DONE]\n\n", sseData)
+
+			resp.Header.Set("Content-Type", "text/event-stream")
+			resp.Header.Set("Cache-Control", "no-cache")
+			resp.Header.Set("Connection", "keep-alive")
+			resp.Body = io.NopCloser(bytes.NewBufferString(sseFormattedData))
+		} else {
+			resp.Body = io.NopCloser(bytes.NewBuffer(rawBody))
+		}
+	}
+
+	return nil
+}
+
+func isModelO1(model string) bool {
```

Comment:
The current PR only supports o1, as o1-mini doesn't seem to support system message which could screw a lot of things. We could investigate in a different issue.